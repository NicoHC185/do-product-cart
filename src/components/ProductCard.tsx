import React from 'react'
import { createContext, CSSProperties, JSX } from 'react'
import { InitialValues, onChangeType, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/product.interface'
import { useProducts } from '../hooks/useProducts'
import styles from '../styles/styles.module.css'

export interface Props {
    product: Product,
    children: (Props: ProductCardHandlers) => JSX.Element,
    className?: string
    style?: CSSProperties
    value?: number
    onChange?: onChangeType
    initialValues?: InitialValues
}

export const ProductContext = createContext({} as ProductContextProps)

const { Provider } = ProductContext

export const ProductCard = ({
    product,
    children,
    className,
    style,
    value,
    onChange,
    initialValues
}: Props) => {

    const { counter, maxCount, increaseBy, isMaxCountReached, reset }
        = useProducts({ product, value, onChange, initialValues })

    return (
        <Provider value={{ counter, maxCount, increaseBy, product }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {children({
                    count: counter,
                    maxCount,
                    product,
                    increaseBy,
                    isMaxCountReached,
                    reset,
                })}
            </div>
        </Provider>
    )
}
