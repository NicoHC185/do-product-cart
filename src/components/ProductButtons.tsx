import React from 'react'

import { CSSProperties, useCallback, useContext } from "react"
import { ProductContext } from "./ProductCard"
import styles from '../styles/styles.module.css'
import { IncreaseBy } from "../interfaces/product.interface"

export interface Props {
    className?: string;
    style?: CSSProperties
}

export const ProductButtons = ({ className, style }: Props) => {

    const { counter, maxCount, increaseBy } = useContext(ProductContext)

    const isMaxReached = useCallback(() => counter === maxCount, [counter, maxCount])

    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <ProductButtonMinus increaseBy={increaseBy} />
            <div className={styles.countLabel}> {counter} </div>
            <ProductButtonAdd increaseBy={increaseBy} disabled={isMaxReached()} />
        </div>
    )
}

interface ButtonProps {
    increaseBy: IncreaseBy,
    disabled?: boolean
}

export const ProductButtonMinus = ({ increaseBy, disabled }: ButtonProps) => {
    return (
        <button
            className={`${styles.buttonMinus} ${disabled && styles.disabled}`}
            onClick={() => !disabled && increaseBy(-1)}
        >
            -
        </button>
    )
}

export const ProductButtonAdd = ({ increaseBy, disabled }: ButtonProps) => {
    return (
        <button
            className={`${styles.buttonAdd} ${disabled && styles.disabled}`}
            onClick={() => !disabled && increaseBy(1)}
        >
            +
        </button>
    )
}

export default ProductButtons