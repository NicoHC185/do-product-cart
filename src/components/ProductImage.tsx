import React from 'react'
import { CSSProperties, useContext, useMemo } from "react"
import { ProductContext } from "./ProductCard"
import noImage from '../assets/no-image.jpg'
import styles from '../styles/styles.module.css'

export interface Props {
    img?: string,
    className?: string
    style?: CSSProperties
}

export const ProductImage = ({ img, className, style }: Props) => {

    const { product } = useContext(ProductContext)

    const imgSrc = useMemo(() => {
        return product?.img ?? img
    }, [product, img])

    return (
        <img
            className={`${styles.productImg} ${className}`}
            style={style}
            src={imgSrc || noImage}
            alt={imgSrc ? 'Product img' : 'no image'}
        />
    )
}

export default ProductImage