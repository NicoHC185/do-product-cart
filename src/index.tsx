import * as React from 'react';
import ProductCard, { ProductButtons, ProductImage, ProductTitle } from './components';

const product = {
  id: '1',
  title: 'Coffe mug - sin imagen',
}

export const Example = () => {
  return (<ProductCard
    product={product}
    initialValues={{
      count: 0,
    }}
  >
    {() => (
      <>
        <ProductImage />
        <ProductTitle />
        <ProductButtons />
      </>
    )}
  </ProductCard>)
};
