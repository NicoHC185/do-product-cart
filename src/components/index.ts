import { ProductButtons } from './ProductButtons'
import { ProductCard as ProductCardHOC } from './ProductCard'
import { ProductCardHOCProps } from '../interfaces/product.interface'
import { ProductImage } from './ProductImage'
import { ProductTitle } from './ProductTitle'

export { default as ProductButtons } from './ProductButtons'
export { default as ProductImage } from './ProductImage'
export { default as ProductTitle } from './ProductTitle'


export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
})


export default ProductCard