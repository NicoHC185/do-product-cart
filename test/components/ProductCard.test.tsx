import React from 'react'
import { ProductCard } from '../../src/components'
import { fireEvent, render } from '@testing-library/react'
import { product1 } from './data/product'

describe('ProductCard', () => {
    test('Debe de mostrar el componente', () => {
        const { container } = render(<ProductCard product={product1}>
            {
                () => (
                    <h1>Product card</h1>
                )
            }
        </ProductCard>)
        expect(container).toMatchSnapshot()
    })

    test('Debe de incrementar el contador', () => {
        const increaseByNumber = 1
        const { asFragment, container } = render(<ProductCard product={product1}>
            {
                ({ count, increaseBy }) => (
                    <>
                        <h1>Product card</h1>
                        <span>{count}</span>
                        <button onClick={() => increaseBy(increaseByNumber)}></button>
                    </>
                )
            }
        </ProductCard>)

        expect(asFragment()).toMatchSnapshot()

        const initialCount = container.querySelector('span')
        expect(initialCount?.textContent).toBe('0')

        fireEvent.click(container.querySelector('button')!)
        const updateCount = container.querySelector('span')
        expect(updateCount?.textContent).toBe(increaseByNumber.toString())
    })
})
