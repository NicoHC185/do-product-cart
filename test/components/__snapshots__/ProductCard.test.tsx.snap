// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Debe de incrementar el contador 1`] = `
<DocumentFragment>
  <div
    class="productCard undefined"
  >
    <h1>
      Product card
    </h1>
    <span>
      0
    </span>
    <button />
  </div>
</DocumentFragment>
`;

exports[`ProductCard Debe de mostrar el componente 1`] = `
<div>
  <div
    class="productCard undefined"
  >
    <h1>
      Product card
    </h1>
  </div>
</div>
`;
